org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,1296 @@
+            // SSL.ConfigJSSDefaultCertAuthCallback(...) on this SSL
+            // PRFileDesc pointer, letting us utilize the same certificate
+            // validation logic that SSLSocket had.
+            if (SSL.ConfigJSSDefaultCertAuthCallback(ssl_fd) == SSL.SECFailure) {
+                throw new RuntimeException("Unable to configure JSSNativeTrustManager on this JSSengine: " + errorText(PR.GetError()));
+            }
+        }
+    }
+
+    public void beginHandshake() {
+        logger.debug("JSSEngine: beginHandshake()");
+
+        // We assume beginHandshake(...) is the entry point for initializing
+        // the buffer. In particular, wrap(...) / unwrap(...) *MUST* call
+        // beginHandshake(...) if ssl_fd == null.
+
+        // ssl_fd == null <-> we've not initialized anything yet.
+        if (ssl_fd == null) {
+            init();
+        }
+
+        // Always, reset the handshake status, using the existing
+        // socket and configuration (which might've been just created).
+        //
+        // Also, NSS's SSL_ResetHandshake takes a parameter "asServer" which
+        // is the opposite of "is_client".
+        if (SSL.ResetHandshake(ssl_fd, !is_client) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to begin handshake: " + errorText(PR.GetError()));
+        }
+    }
+
+    public void closeInbound() {
+        logger.debug("JSSEngine: closeInbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_RCV);
+        is_inbound_closed = true;
+    }
+
+    public void closeOutbound() {
+        logger.debug("JSSEngine: closeOutbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_SEND);
+        is_outbound_closed = true;
+    }
+
+    public String getHostname() {
+        return hostname;
+    }
+
+    public Runnable getDelegatedTask() {
+        logger.debug("JSSEngine: getDelegatedTask()");
+
+        // We fake being a non-blocking SSLEngine. In particular, we never
+        // export tasks as delegated tasks (e.g., OCSP checking), so this
+        // method will always return null.
+
+        return null;
+    }
+
+    private void queryEnabledCipherSuites() {
+        logger.debug("JSSEngine: queryEnabledCipherSuites()");
+        ArrayList<SSLCipher> enabledCiphers = new ArrayList<SSLCipher>();
+
+        for (SSLCipher cipher : SSLCipher.values()) {
+            try {
+                if (SSL.CipherPrefGet(ssl_fd, cipher.getID())) {
+                    enabledCiphers.add(cipher);
+                }
+            } catch (Exception e) {
+                // Do nothing -- this shouldn't happen as SSLCipher should be
+                // synced with NSS. However, we'll just log this exception as
+                // a warning. At worst we fail to report that a cipher suite is
+                // enabled.
+                logger.warn("Unable to get the value of cipher: " + cipher.name() + " (" + cipher.getID() + "): " + e.getMessage());
```

Comment:
In general I'd rather not ignore an exception just because it "shouldn't happen". I suppose the `SSL.CipherPrefGet()` is a read operation. If such an operation fails doesn't it indicate a probably worse underlying issue? In that case we might want to let the caller knows about such problem.

Reply:
Not necessarily. If it fails, it could just be that JSS development has passed NSS development, or visa-versa. There's no real risk (outside of not being able to use a particular cipher) here. Otherwise, when NSS rebases and introduces a new cipher, we'd _always_ have to patch `SSLCipher`, if they introduce a new underlying cipher.

Same thing happens in reverse: if we want to build JSS so that it _could_ take advantage of a newer cipher when it is say, introduced in upstream NSS -- but prior to its introduction on downstream NSS -- it'd be beneficial to have a larger `SSLCipher` list and then when it gets introduced, it is possible to pick it up with no changes to JSS (just a new NSS version downstream).

Reply:
> so ideally JSS should never be ahead of NSS in terms of cipher support for the corresponding platform.

I disagree. JSS used to be shipped by Mozilla, but it isn't any more. JSS is really the only high-level bindings to NSS that exist. JSS should strive to be independent from NSS releases and work with as many NSS versions as possible. In particular, JSS (at any given release) should match what NSS master branch has. This lets it know about newer algorithms than what the version of NSS it was _compiled_ with has, _if_ the _runtime_ NSS also supports those cipher suites.

In reality, the situation is much more complex. NSS "knows" about a bunch of algorithms. Including really stupid ones like `TLS_NULL_WITH_NULL_NULL`. However, any attempt to set them reports an error (getting them doesn't). Why? Because they're really stupid algorithms and nobody should be using them.

IMO, we should gracefully handle this: SSLCipher knows about _every_ cipher suite that NSS master does. If NSS, at runtime, knows about the cipher, we can use it. Nothing prevents us from making that guarantee.

---

We also don't ship NSS either. We can't ensure JSS and NSS are always updated in lock-step. They're two separate packages after all! At best, we can try to be compatible with as many NSS versions as possible. And sometimes, things will break. Take: 3.47 <= NSS < 3.51 for instance: it doesn't work with JSS master's KBKDF/CMAC support. That's OK. Note it, and move on. But that doesn't stop nearly all of JSS from working on that platform.

--- 

Additionally, this particular call really doesn't matter. This happens when:

 1. No SSLPreferences have been applied,
 2. No call to enable specific cipher suites have been made,
 3. Someone asks what is enabled (not, I might add, what is _supported_). 

So there is no risk to ignoring this here. At worst, we report that a cipher suite enabled by NSS isn't enabled from the PoV of SSLEngine. That's fine. We're probably expecting someone to set their own list of ciphers afterwards anyways. And then we'll disable all cipher suites anyways, and re-enable only the ones they ask for.

Reply:
Here's the code:

```c
/* return the user preference for this suite */
SECStatus
ssl3_CipherPrefGetDefault(ssl3CipherSuite which, PRBool *enabled)
{
    const ssl3CipherSuiteCfg *suite;
    PRBool pref;
    SECStatus rv;

    suite = ssl_LookupCipherSuiteCfg(which, cipherSuites);
    if (suite) {
        pref = suite->enabled;
        rv = SECSuccess;
    } else {
        pref = SSL_NOT_ALLOWED;
        rv = SECFailure; /* err code was set by Lookup. */
    }
    *enabled = pref;
    return rv;
}
```

`ssl3_CipherPrefGetDefault` fails if and only if `ssl_LookupCipherSuiteCfg` fails. 

This fails only in the case where the cipher suite is unknown:

```c
/* Find the cipher configuration struct associate with suite */
/* XXX This does a linear search.  A binary search would be better. */
static ssl3CipherSuiteCfg *
ssl_LookupCipherSuiteCfgMutable(ssl3CipherSuite suite,
                                ssl3CipherSuiteCfg *suites)
{
    int i;

    for (i = 0; i < ssl_V3_SUITES_IMPLEMENTED; i++) {
        if (suites[i].cipher_suite == suite)
            return &suites[i];
    }
    /* return NULL and let the caller handle it.  */
    PORT_SetError(SSL_ERROR_UNKNOWN_CIPHER_SUITE);
    return NULL;
}

const ssl3CipherSuiteCfg *
ssl_LookupCipherSuiteCfg(ssl3CipherSuite suite, const ssl3CipherSuiteCfg *suites)
{
    return ssl_LookupCipherSuiteCfgMutable(suite,
                                           CONST_CAST(ssl3CipherSuiteCfg, suites));
}
```

In all cases, this only happens if the cipher suite is unknown. Since we use [`SSLCipher`](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/ssl/SSLCipher.html) enum [internally to `JSSEngine`](https://github.com/cipherboy/jss/blob/ssl-engine/org/mozilla/jss/ssl/javax/JSSEngine.java#L119-L125), this means that JSS and NSS are out of sync. That's not really a bug and not something the user should really be concerned about. 

Its at most, a point of information. Hence, a logged debug message.

---

Also, this is only warned when the user doesn't set cipher suites ... but wants to know what is enabled currently. Since everything that is enabled by default in NSS by necessity complies with `crypto-policies` and NSS claims it can negotiate it, I don't think this is a problem.